src/formatting_stack/strategies.clj

```diff
@@ -36,12 +36,16 @@
                        (map git-status/remove-deletion-markers)
                        (impl/absolutize git-command)
                        (set))
-          tracked (->> (impl/file-entries git-command "ls-files" "--full-name")
-                       (impl/absolutize git-command)
-                       (remove deleted))
-          untracked (->> (impl/file-entries git-command "ls-files" "--full-name" "--others" "--exclude-standard")
-                         (impl/absolutize git-command)
-                         (remove deleted))]
+          ;; DRY the use of the `--full-name` option which is especially important:
+          ls-files (speced/fn [^list? args]
+                     (->> args
+                          (cons "--full-name")
+                          (cons "ls-files")
+                          (apply impl/file-entries git-command)
+                          (impl/absolutize git-command)
+                          (remove deleted)))
+          tracked (ls-files ())
+          untracked (ls-files '("--others" "--exclude-standard"))]
```

Comment:
using `& args` would result in a `list?` for `args` too, and make the api more elegant; 

i.e.
```clj
(ls-files)
(ls-files "--others" "--exclude")
```

Reply:
It's a good suggestion, however two considerations apply:

* I prefer that `list?` is part of the explicit contract, and not something that can go away unintentionally (and perhaps break due to vector<->list differences)
* `& args` implies `apply` will be used internally, which performs worse

Performance tends to be a good 'decider' in case of doubt. I also tend to favor performant code regardless of whether is needed, simply as a habit that can pay off in unsuspected places.

Reply:
> I prefer that list? is part of the explicit contract, and not something that can go away unintentionally (and perhaps break due to vector<->list differences)

You can still spec `args`, explicitly:
```clj
(speced/fn [& ^list? args] ...)
```

> & args implies apply will be used internally, which performs worse

This fn is not run in a tight loop or anything, so it's no need for a premature optimalisation.

Reply:
> This fn is not run in a tight loop or anything, so it's no need for a premature optimalisation.

Over the years I've found a no-broken-windows approach to performance is cheap and effective enough: if performance can be gained by applying simple habits, then it should.

The fixed cost of `(apply f)` is 2x as slow as that of `(f)`. For an input large enough (here, for example: a large monorepo), that 2x often becomes noticeable.

Anyway the point I'm defending is precisely the opposite of fine-grained reasoning: in case of doubt (especially given two equally defensible choices), one can use an _objective measurement_ as a decider (and guide for everyday programming).

---

It's a similar reasoning I use to design speced.def (with spec->type inference), pay attention to compiler/JVM flags in our libs, etc. All those can be seen as 'irrelevant' but they do add up and free one of having to chase false leads when a real problem arises.

Reply:
> Over the years I've found a no-broken-windows approach to performance is cheap and effective enough: if performance can be gained by applying simple habits, then it should.

If _tangible_ performance can be gained [...] Sure. In this case it's sub nanosecond. Which, in a higher order language, is not tangible whatsoever.

> For an input large enough

The api in your case has 0..2 arguments, a large monorepo affect that. Keep in mind that using `apply` is not needed at all.

Reply:
Please don't get past of what the no-broken-windows principle stands for: one follows habits, no matter what, no exceptions, because breaking them opens a slippery slope. It's that slippery-slope what I'm afraid of: by setting concrete precedents, future changes to PRs can similarly disregard performance. Individual reasonings from the past can stop holding true, in face of code changes, concurrent PRs, a diversity of authors, unforeseen use cases, etc.

So please let's not break that window?

And also, please observe that the alternative offers no tangible benefits. One is erasing a ` ()` and adding a ` &`. Performance worsens and nothing else changes. I'm not really _against_ your proposal, but we should be able to have an objective decider when in doubt.

Let's take this constructively as an opportunity to reflect what a code review is _for_? It rarely is for altering a diff, character by character, until it looks exactly as a reviewer would have authored himself.

Interestingly, the long-term (and original) goal of formatting-stack is precisely elliminating these churn-y code reviews: if a PR passes sufficiently smart checks (including test coverage analysis, which to be fair I have barely leveraged) and satisfies business requirements, it should be essentially fine.

> Keep in mind that using apply is not needed at all.

`&` implies `apply` at the compiler level, it's what is being invoked under the hood and why any defn with `&` is slower.

While one can reason whether some optimization would be irrelevant, one can also observe that _fine-grained reasoning doesn't scale_, which is another key driver for formatting-stack (and speced.def). By using heavy linting, assertions, etc one can catch things that otherwise an 'outdated reasoning' would have not caught.

Reply:
It is clear that this discussion is not going anywhere. I do not share your personal views expressed here, but given the impact of the change (better said; the lack thereof), I'll leave this be.